Add a new prometheus metrics in m.qp.middleware.metrics/adjust

appleby opened the pull request:
### Description

Context: https://metaboat.slack.com/archives/C0645JP1W81/p1728471378145769

Add a new prometheus counter to count errors in `metabase.query-processor.middleware.metrics/adjust`.

### How to verify

* Set `MB_PROMETHEUS_SERVER_PORT` and then check the `/metrics` endpoint, should see logs like

```
# HELP metabase_query_processor_metrics_total Number metrics processed by the query processor.
# TYPE metabase_query_processor_metrics_total counter
metabase_query_processor_metrics_total 0.0
# HELP metabase_query_processor_metrics_errors_total Number of errors when processing metrics.
# TYPE metabase_query_processor_metrics_errors_total counter
metabase_query_processor_metrics_errors_total 0.0
# HELP metabase_query_processor_metrics_created Number metrics processed by the query processor.
# TYPE metabase_query_processor_metrics_created gauge
metabase_query_processor_metrics_created 1.728947783982E9
# HELP metabase_query_processor_metrics_errors_created Number of errors when processing metrics.
# TYPE metabase_query_processor_metrics_errors_created gauge
metabase_query_processor_metrics_errors_created 1.728947783982E9
```

* Or see [dpsutton's comment](https://github.com/metabase/metabase/pull/48547#issuecomment-2405875446) about testing from the repl
* See included tests

### Checklist

- [x] Tests have been added/updated to cover changes in this PR

appleby committed:
Add new promethues counter for :metabase-metrics/adjust-errors

To record errors in metabase.query-processor.middleware.metrics/adjust

appleby committed:
Add prometheus metrics test to metrics middleware tests

appleby commented on src/metabase/analytics/prometheus.clj:
```diff
@@ -194,11 +194,13 @@
    (prometheus/counter :metabase-email/message-errors
-                       {:description (trs "Number of errors when sending emails.")})])
+                       {:description (trs "Number of errors when sending emails.")})
+   (prometheus/counter :metabase-metrics/adjust-errors
```
Any naming conventions I should follow for the metric name here?

appleby replied:
Do we want to also count the total number of adjustments attempted? I included a counter for it in a first pass, but backed it out since I don't have a good sense of how often `adjust` will wind up getting called.

metamben replied:
I suppose the interesting thing to know would be how often it processed a query with metrics (i.e., had to change the query) and how often the invariant that no `:metric` should be in the result failed.

appleby replied:
I added the "total adjustments attempted" counter and used the suggested names. Easy to change if @dpsutton has other ideas about naming.

appleby commented on src/metabase/query_processor/middleware/metrics.clj:
```diff
@@ -254,5 +255,6 @@
       (update query :stages #(adjust-metric-stages query nil %))
       (when-let [metric (lib.util.match/match-one <>
                           [:metric _ _] &match)]
+        (prometheus/inc! :metabase-metrics/adjust-errors)
```
I guess the `match-one` means we're only counting the # calls that result in at least one failure, not the total number of `:metrics` we failed to adjust.

appleby replied:
Just noticed there are several places in this ns that throw an exception for incompatible or malformed metrics. Maybe we should count those as well? Hmm, should the metric name be generalized to something like `:metabase-metrics/errors` or should a separate counter be created for each location that throws? 🤔

appleby committed:
Also inc :metabase-metrics/adjust-errors when throwing an exception

appleby committed:
Make splice-compatible-metrics private

It doesn't seem to be used outside of metabase.query-processor.middleware.metrics.

appleby commented on src/metabase/query_processor/middleware/metrics.clj:
```diff
@@ -114,7 +123,7 @@
     (lib.util/update-query-stage query-with-joins agg-stage-index add-join-aliases source-field->join-alias)))
 
-(defn splice-compatible-metrics
+(defn- splice-compatible-metrics
```
Unrelated, but AFAICT nothing outside this namespace was using it, so I made it private.

appleby commented on test/metabase/query_processor/middleware/metrics_test.clj:
```diff
@@ -80,6 +81,17 @@
                        :aggregation [[:avg {} [:field {} (meta/id :products :rating)]]]}]}
             (adjust query)))))
 
+(deftest adjust-prometheus-metric-test
```
This probably needs more test cases now that we inc the counter in multiple places

appleby committed:
Add test for :metabase-metrics/adjust-errors when exception is thrown

appleby committed:
Move inc-and-throw! helper into metabase.analytics.prometheus

appleby commented on src/metabase/analytics/prometheus.clj:
```diff
@@ -258,6 +260,12 @@
   [metric]
   (some-> system .-registry metric prometheus/inc))
 
+(defn inc-and-throw!
```
Ambivalent about this tiny wrapper. On the one hand, if we're going to use prometheus metrics more going forward, it will presumably be common to want to increment an error counter and throw an exception together. At the same time, it's not hard to wrap these two calls in a `do` block, and small wrappers like this can complicate linters or other static analysis tools that want to check, e.g., every place an exception is thrown.

metamben replied:
I like this. We might want to teach the linter that this function always throws.

dpsutton replied:
reading a bit more, i suspect we don't want an API like this. We have thoughtful error handling in certain places and we probably don't want to tie our hands with an overly simplistic api like this?

appleby replied:
I was also worried about `inc-and-throw!` cluttering up the stacktrace, so this is the first thing I checked, but it seems that Clojure attaches the stack at the point the `Throwable` is constructed, not at the point it is thrown, so the top frame will still be the callers frame.

```Clojure
dev> (defn call-inc-and-throw! []
       (prometheus/inc-and-throw! :foo (Exception. "foo")))
#'dev/call-inc-and-throw!
dev> (call-inc-and-throw!)
Execution error at dev/call-inc-and-throw! (REPL:1997).
foo
```

```
1. Unhandled java.lang.Exception
   foo

                      REPL: 1997  dev/call-inc-and-throw!
                      REPL: 1996  dev/call-inc-and-throw!
                      REPL: 1999  dev/eval209000
                      REPL: 1999  dev/eval209000
             Compiler.java: 7700  clojure.lang.Compiler/eval
    interruptible_eval.clj:  106  nrepl.middleware.interruptible-eval/evaluator/run/fn
    interruptible_eval.clj:  101  nrepl.middleware.interruptible-eval/evaluator/run
               session.clj:  229  nrepl.middleware.session/session-exec/session-loop
        SessionThread.java:   21  nrepl.SessionThread/run
```

The motivation here was to be an easy way to convert any `throw` to an `inc-and-throw!` and get error counting _without_ disturbing any thoughtful error handling surrounding it, and, in particular, to reduce diff in the common situation where the `throw` is in one branch of some conditional, where you would otherwise need to wrap it in `do` to insert the `inc!`. But as @metamben [rightly pointed out](https://github.com/metabase/metabase/pull/48598#discussion_r1799965415), most of the time you just want a single `try`/`catch` at some entry point to the module that's counting errors, so this wrapper is now even less useful than when I added it, and I'm happy to remove it if there are objections.

appleby replied:
I went ahead and removed `inc-and-throw!` for the following reasons

1. Opinions are divided (even I am not fully convinced tbh). When in doubt THROW IT OUT.
2. Recent updates to this PR mean there is now only a single call site, so value is questionable.
3. Easy to add back in the future if it proves useful other places, and API inertia means it's easier to add something to an API than it is to take it away later if we change our minds about it.

metamben commented on test/metabase/analytics/prometheus_test.clj:
```diff
@@ -172,5 +172,8 @@
-      (prometheus/inc! :metabase-email/messages)
-      (is (< 0 (-> system :registry :metabase-email/messages ops/read-value))))))
+      (let [read-value #(-> system :registry % ops/read-value)]
+        (doseq [metric [:metabase-email/messages :metabase-metrics/adjust-errors]]
```
If it's necessary to test this for every metric we introduce, then this is not the best way to go about it: we should have a way to figure out automatically what has to be tested (like we do with serdes stuff). But I suppose we don't have to test every metric.

appleby replied:
Yeah this is probably not needed, I will roll this back. But this also reminds me that I should now add a test for `inc-and-throw!` (assuming @dpsutton does not object to adding that wrapper to the lib).

metamben replied:
The tricky part with `inc-and-throw!` is to make sure we don't count the same error several times. I mean the other pattern we could follow to have a single catch in `adjust` that increments and re-throws. (This was my initial instinct.)

appleby replied:
Sure, could be nice to have the metrics recorded in a single place in `adjust`. This also has the added benefit that if _any_ exception is thrown, we record the error, even if it wasn't thrown directly by the metrics middleware, which I think is what we want. I will make this change.

metamben reviewed (commented):
I suggest we wait a little to let @dpsutton chime in.

appleby committed:
Add inc-and-throw! test to prometheus tests

appleby committed:
Rename inc-test to inc!-test and use approx= and metric-value helpers

appleby committed:
Add try/catch to metrics adjust and record metric errors in a single place

Record promethues metrics in a single place in metabase.query-processor.middleware.metrics and ensure we record
failures for any exception throw.

appleby committed:
Improve testing description string

appleby committed:
Rename :metabase-metrics/adjust-errors to :metabase-query-processor/metrics-errors

appleby committed:
Move metrics adjustment try/catch into a new helper

This ensures we count all exceptions generated by adjust-metric-stages.

appleby committed:
Add new prometheus metric to count number of MB metrics seen by the QP

This new metric is intended to count the number of metrics that the query processor metrics middleware sees and
attempts to adjust.

appleby committed:
Add more metrics metrics tests

- Ensure that queries with no metrics do not incremement either counter
- Ensure that successful calls to adjust do not increment the error counter
- Ensure that exceptions throw from other libs also increment the error counter

appleby committed:
Rename test to adjust-prometheus-metrics-test and add helper to reduce code dup

appleby committed:
If no metrics are found immediately return query in metrics middleware adjust

Don't attempt to adjust query if we don't find any :metric clauses.

appleby commented on src/metabase/query_processor/middleware/metrics.clj:
```diff
@@ -245,14 +258,18 @@
-                          [:metric _ _] &match)]
-        (log/warn "Failed to replace metric"
-                  (pr-str {:metric metric}))))))
+  (if-not (match-one-metric query)
```
@metamben N.B. this change, we no longer call `adjust-metric-stages` if we don't find a `:metric` clause (otherwise, we update the "total adjustments" counter and attempt to adjust). Looks safe to me but let me know if not.

metamben replied:
I expect this should be a no-op if the query isn't consuming metrics. 👍

metamben commented on src/metabase/analytics/prometheus.clj:
```diff
@@ -194,11 +194,15 @@
-                       {:description (trs "Number of errors when sending emails.")})])
+                       {:description (trs "Number of errors when sending emails.")})
+   (prometheus/counter :metabase-query-processor/metrics
+                       {:description (trs "Number metrics processed by the query processor.")})
```
```suggestion
                       {:description (trs "Number of queries consuming metrics processed by the query processor.")})
```

metamben commented on src/metabase/query_processor/middleware/metrics.clj:
```diff
@@ -230,6 +231,18 @@
       :else
       expanded-stages)))
 
+(defn- adjust-metric-stages-counting-errors
```
What about sg. like `adjust-metric-stages-counting-failure` or `adjust-metric-stages-booking-failure`?
This name suggests me that it will count some errors during processing.

As noted below, I'm not sure this function is needed at all.

appleby replied:
I also do not like the name or even the existence of this function. See response below, but I will remove this.

metamben commented on src/metabase/query_processor/middleware/metrics.clj:
```diff
@@ -230,6 +231,18 @@
+(defn- match-one-metric
+  [query]
+  (lib.util.match/match-one query
+    [:metric _ _] &match))
```
What about `references-metrics?` (and calling `boolean` on the result of `match-one`)?

appleby replied:
The return value is used in the `when-let` when checking to see if the metric replacement failed.

We could do without this function of course, but I wanted to make it clear that the `if-not` and `when-let` were checking for the same condition.

metamben commented on src/metabase/query_processor/middleware/metrics.clj:
```diff
@@ -245,14 +258,18 @@
-                  (pr-str {:metric metric}))))))
+  (if-not (match-one-metric query)
+    query
+    (do
```
Why not have the `try ... catch` here?

appleby replied:
That is what I did initially, but then I considered your comment [here](https://github.com/metabase/metabase/pull/48598#discussion_r1799965415) about being careful not to double-count errors and realized there is some very small chance that `prometheus/inc!` or (even less likely) `log/warn` could throw, in which case the error could be double-counted, but I'm probably being overly paranoid, especially since `inc!` would have to throw _after_ updating the counter to be double counted, which seems extremely unlikely. Per your comment above I don't love `adjust-metric-stages-counting-errors` either, so maybe this is a case of the cure being worse than the disease. I will update this to move the `try` / `catch` block here.

metamben commented on test/metabase/analytics/prometheus_test.clj:
```diff
@@ -164,7 +164,22 @@
+  ([expected actual]
+   (approx= expected actual 0.001))
+  ([expected actual epsilon]
+   (< (- expected epsilon) actual (+ expected epsilon))))
```
Works for me, but I think the canonical form is
```suggestion
   (< (abs (- actual expected)) epsilon)))
```

metamben commented on test/metabase/query_processor/middleware/metrics_test.clj:
```diff
@@ -60,6 +61,72 @@
+                             (query-fn mp source-metric)
+                             (lib/query mp source-metric))
+        calls              (atom nil)]
+    (with-redefs [prometheus/inc! #(swap! calls conj %)]
```
Why not count directly with sg. like
```suggestion
    (with-redefs [prometheus/inc! #(swap! calls update % (fnil inc 0))]
```

appleby replied:
Tbh I copied this idiom from the existing prometheus-metrics-counting tests in `metabase.email-test/send-message!-test`, but agree that counting directly seems better.

appleby committed:
Appease cljfmt

dpsutton commented:
One thing i notice about the tests is that we're spending ~80 lines writing tests to see if a number gets incremented or not. But we can just instrument the existing tests. Here I'm grabbing the before and after values. Perhaps we could just instrument one or two of these?

```clojure
metrics-test=> (let [f! #(-> #'metabase.analytics.prometheus/system var-get :registry :metabase-query-processor/metrics iapetos.core/value)
                     e! #(-> #'metabase.analytics.prometheus/system var-get :registry :metabase-query-processor/metrics-errors iapetos.core/value)]
                 (doseq [[s v] (ns-publics *ns*)
                         :when (-> v meta test)]
                   (let [before (f!)
                         before-e (e!)]
                     (try (v) (catch Exception _e))
                     (println s ": incremented " (long (- (f!) before)) " errors: " (long (- (e!) before-e))))))
custom-aggregation-test : incremented  1  errors:  0
metric-question-on-native-model-test : incremented  1  errors:  0
e2e-source-card-test : incremented  0  errors:  0
adjust-filter-test : incremented  1  errors:  0
metric-question-on-model-based-on-model-test : incremented  1  errors:  0
model-based-metric-with-implicit-join-test : incremented  1  errors:  0
available-metrics-test : incremented  0  errors:  0
metric-with-explicit-join-test : incremented  1  errors:  0
adjust-aggregation-metric-ref-test : incremented  1  errors:  0
override-source-names-test : incremented  1  errors:  0
no-metric-should-result-in-exact-same-query : incremented  0  errors:  0
metric-in-offset-test : incremented  1  errors:  0
metric-with-implicit-join-test : incremented  3  errors:  0
metric-question-on-multi-stage-model-test : incremented  1  errors:  0
metric-question-on-aggregate-column-model-test : incremented  1  errors:  0
execute-single-stage-metric : incremented  3  errors:  0
adjust-mixed-multi-source-test : incremented  2  errors:  0
default-metric-names-test : incremented  1  errors:  0
adjust-basic-source-table-test : incremented  1  errors:  0
filtered-metric-test : incremented  1  errors:  0
adjust-aggregation-metric-ordering-test : incremented  1  errors:  0
e2e-source-metric-results-test : incremented  1  errors:  0
joined-question-based-on-metric-based-on-metric-based-on-metric-test : incremented  2  errors:  0
adjust-prometheus-metrics-test : incremented  0  errors:  0
include-source-names-test : incremented  1  errors:  0
metric-with-nested-segments-test : incremented  2  errors:  0
multiple-source-metrics-with-implicit-join-test : incremented  1  errors:  0
adjust : incremented  0  errors:  0
question-based-on-metric-based-on-metric-based-on-metric-test : incremented  3  errors:  0
adjust-expression-test : incremented  1  errors:  0
maintain-aggregation-refs-test : incremented  2  errors:  0
adjust-join-test : incremented  1  errors:  0
adjust-basic-source-metric-test : incremented  1  errors:  0
e2e-source-table-results-test : incremented  1  errors:  0
metric-question-on-custom-column-model-test : incremented  1  errors:  0
model-based-metric-use-test : incremented  1  errors:  0
adjust-expression-name-collision-test : incremented  1  errors:  0
expand-macros-in-nested-queries-test : incremented  2  errors:  0
```

appleby committed:
Improve metric description

Co-authored-by: metamben <103100869+metamben@users.noreply.github.com>

appleby committed:
Improve epsilon test in approx=

Co-authored-by: metamben <103100869+metamben@users.noreply.github.com>

appleby commented:
> One thing i notice about the tests is that we're spending ~80 lines writing tests to see if a number gets incremented or not. But we can just instrument the existing tests. Here I'm grabbing the before and after values. Perhaps we could just instrument one or two of these?

I see two separate (and valid!) parts to this:

1. Why not read the counter values directly, rather than mocking out calls to `prometheus/inc!`.
2. Why not instrument existing tests, rather than write (tedious) bespoke hand-crafted artisanal tests

For (1), reading the counter values was my first instinct as well, possibly adding some helper function/macro to capture the before/after values or else reset the counters (or maybe just use existing `with-prometheus-system!` in the prometheus tests), but then I stumbled on [the discussion](https://github.com/metabase/metabase/pull/31333#issuecomment-1639803104) between yourself and devurandom in #31333, and the arguments there by devurandom about keeping testing concerns separate seemed reasonable, so I followed that approach.

For (2), I also considered this! Possibly adding metadata to existing tests to filter them for inclusion/instrumentation. In the end I was worried this would be brittle as the tests could break if anyone updates the existing tests to run more queries, etc. It wouldn't be hard to go back and update the expectation to match, but I guess it could be mildly annoying / confusing if you aren't aware the prometheus tests exist, because typically if you modify some test, you also modified some code as well, so now you need to convince yourself that the new values are correct (not hard but still).

appleby commented:
Moving this back to draft to respond to PR feedback

appleby committed:
Remove adjust-metric-stages-counting-errors

This function was an attempt to avoid double-counting errors, but probably not worth the effort.

See

https://github.com/metabase/metabase/pull/48598#discussion_r1800197964
https://github.com/metabase/metabase/pull/48598#discussion_r1800200421

appleby committed:
Count metrics directly when mocking prometheus/inc!

Based on PR feedback here:

https://github.com/metabase/metabase/pull/48598#discussion_r1800207674

appleby committed:
Remove prometheus/inc-and-throw!

Value prop was always questionable, but now that it's only called from a single place, just remove it and inline the
call to prometheus/inc!. Easy to add back later if it proves generally useful.

PR discussion thread

https://github.com/metabase/metabase/pull/48598#discussion_r1799728596

appleby commented:
> One thing i notice about the tests is that we're spending ~80 lines writing tests to see if a number gets incremented or not. But we can just instrument the existing tests. Here I'm grabbing the before and after values. Perhaps we could just instrument one or two of these?

Totally agree with the underlying concern here though, which is that it's not ideal if we have to add 80ish lines of bespoke tests every time we add prometheus metrics to some module, and it would be better if we had some low-cost / low-friction way to add such tests to an existing module.

metamben reviewed (approved):
LGTM. I have some suggestion for your consideration, feel free to ignore/postpone/whatever them.

appleby committed:
Rename match-one-metric to find-first-metric

https://github.com/metabase/metabase/pull/48598#discussion_r1803321025

appleby committed:
Rename :metabase-query-processor/metrics-errors to metric-errors

https://github.com/metabase/metabase/pull/48598#discussion_r1803313818

appleby committed:
Throw an exception rather than logging a warning if we fail to adjust a metric

https://github.com/metabase/metabase/pull/48598#discussion_r1803328482

appleby commented on src/metabase/analytics/prometheus.clj:
```diff
@@ -194,11 +194,15 @@
+                       {:description (trs "Number of errors when sending emails.")})
+   (prometheus/counter :metabase-query-processor/metrics
+                       {:description (trs "Number of queries consuming metrics processed by the query processor.")})
+   (prometheus/counter :metabase-query-processor/metrics-errors
```
Done. This is great attention to detail, and the kind of nit I love to see picked!

Fwiw, I have no idea why, but in isolation "metrics-errors" feels / sounds more natural to me than "metric-errors" (though both are fine). But "messages-errors" is definitely awkward, and consistency in naming is good.

metamben replied:
Well, you are the native speaker, don't listen to me in such matters. (Until we start writing code in Hungarian.)

appleby commented on src/metabase/query_processor/middleware/metrics.clj:
```diff
@@ -245,14 +251,24 @@
+            (update query :stages #(adjust-metric-stages query nil %))
+            (when-let [metric (match-one-metric <>)]
+              ;; There is a small chance we could wind up double-counting errors here if prometheus/inc! or log/warn
+              ;; throw an exception, but it seems unlikely enough to not warrant complications to avoid it.
```
**done**
